contrib/platform/src/com/sun/jna/platform/unix/LibCUtil.java

```diff
@@ -0,0 +1,140 @@
+/* Copyright (c) 2020 Daniel Widdis, All Rights Reserved
+ *
+ * The contents of this file is dual-licensed under 2
+ * alternative Open Source/Free licenses: LGPL 2.1 or later and
+ * Apache License 2.0. (starting with JNA version 4.0.0).
+ *
+ * You can freely decide which license you want to apply to
+ * the project.
+ *
+ * You may obtain a copy of the LGPL License at:
+ *
+ * http://www.gnu.org/licenses/licenses.html
+ *
+ * A copy is also included in the downloadable source code package
+ * containing JNA, in file "LGPL2.1".
+ *
+ * You may obtain a copy of the Apache License at:
+ *
+ * http://www.apache.org/licenses/
+ *
+ * A copy is also included in the downloadable source code package
+ * containing JNA, in file "AL2.0".
+ */
+package com.sun.jna.platform.unix;
+
+import com.sun.jna.Function;
+import com.sun.jna.Native;
+import com.sun.jna.NativeLibrary;
+import com.sun.jna.Pointer;
+
+/**
+ * Utility class supporting variable-width types in the C Library
+ */
+public class LibCUtil {
+
+    private static final NativeLibrary LIBC = NativeLibrary.getInstance("c");
+
+    /** Size of a native <code>off_t</code> type, in bytes. */
+    public static final int OFF_T_SIZE;
+    static {
+        // Observations shows, that without LFS, on linux, solaris, aix, mac OS and
+        // windows (apart from cygwin32) sizeof(off_t) == sizeof(long)
+        int size = Native.LONG_SIZE;
+        // On 64-bit, off_t is 64 bit, otherwise test compiler flags that would create
+        // 64-bit versions of off_t functions
+        if (size < 8) {
+            try {
+                LIBC.getFunction("mmap64", Function.THROW_LAST_ERROR);
+                // on 32-bit, mmap64 only exists when off_t is 64-bit
+                size = 8;
```

Comment:
OK, I agree with you. Documenting my deep dive of the code here for future reference:

Consistently across most *nix, `__off_t` is mapped to `long` and `__off64_t` to `long long`.  These are mapped to `off_t` in [mman.h](https://sites.uclouvain.be/SystInfo/usr/include/sys/mman.h.html):
```
#ifndef __off_t_defined
# ifndef __USE_FILE_OFFSET64
typedef __off_t off_t;
# else
typedef __off64_t off_t;
# endif
# define __off_t_defined
#endif
```
So we only really care about the 32-bit OS case detecting 64-bit `off_t`, which happens only in the case of `__USE_FILE_OFFSET64`. 

Later, the `mmap` mapping also uses  `__USE_FILE_OFFSET64`, which actually uses the base type mapping (`__off_t` or `__off64_t`) without reference to the userspace `off_t`:
```
#ifndef __USE_FILE_OFFSET64
extern void *mmap (void *__addr, size_t __len, int __prot,
                   int __flags, int __fd, __off_t __offset) __THROW;
#else
# ifdef __REDIRECT_NTH
extern void * __REDIRECT_NTH (mmap,
                              (void *__addr, size_t __len, int __prot,
                               int __flags, int __fd, __off64_t __offset),
                              mmap64);
# else
#  define mmap mmap64
# endif
#endif
#ifdef __USE_LARGEFILE64
extern void *mmap64 (void *__addr, size_t __len, int __prot,
                     int __flags, int __fd, __off64_t __offset) __THROW;
#endif
```
From this latter mapping we see `mmap64` only exists if either `__USE_FILE_OFFSET64` (corresponding directly to the `off_t` bit width we care about) or when `__USE_LARGEFILE64` is defined.  

When `__USE_FILE_OFFSET64` is defined we know from above that `off_t` is 64-bit. The only case of concern would be if `__USE_LARGEFILE64` is defined (giving us `mmap64`) but `__USE_FILE_OFFSET64` is not (giving us 32-bit `off_t`).

Two things can cause `__USE_LARGEFILE64` to be defined. See [features.h](https://code.woboq.org/qt5/include/features.h.html).

One way is the user-controled `-D_FILE_OFFSET_BITS = 64` which causes `__USE_FILE_OFFSET64` to be defined at line 363:
```
#if defined _FILE_OFFSET_BITS && _FILE_OFFSET_BITS == 64
# define __USE_FILE_OFFSET64	1
#endif
```
and later at line 431 we see this define causes `__USE_LARGEFILE_64` to be defined.
```
/* If we don't have __REDIRECT, prototypes will be missing if
   __USE_FILE_OFFSET64 but not __USE_LARGEFILE[64]. */
# if defined __USE_FILE_OFFSET64 && !defined __REDIRECT
#  define __USE_LARGEFILE	1
#  define __USE_LARGEFILE64	1
# endif
```
So we're safe there. The other way to get `__USE_LARGEFILE64` defined is at line 359:
```
#ifdef _LARGEFILE64_SOURCE
# define __USE_LARGEFILE64	1
#endif
```
Here, the top of the file documents the LFS functionality:
`_LARGEFILE64_SOURCE	Additional functionality from LFS for large files.`
and it is defined only if the user specifies `_GNU_SOURCE` (see line 190):
```
/* If _GNU_SOURCE was defined by the user, turn on all the other features.  */
#ifdef _GNU_SOURCE
... other defines ...
# undef	 _LARGEFILE64_SOURCE
# define _LARGEFILE64_SOURCE	1
... other defines ...
#endif
```

All the above is [documented here](https://www.gnu.org/software/libc/manual/html_node/Feature-Test-Macros.html), noting:
> Macro: _LARGEFILE64_SOURCE
> If you define this macro an additional set of functions is made available 

and
> Macro: _FILE_OFFSET_BITS
> This macro determines which file system interface shall be used, one replacing the other. Whereas _LARGEFILE64_SOURCE makes the 64 bit interface available as an additional interface, _FILE_OFFSET_BITS allows the 64 bit interface to replace the old interface.

So, this does indicate a case where my code would fail: 
1. a user did NOT specify -D_FILE_OFFSET_BITS=64 at C library compile time
2. a user did specify either `-D_LARGEFILE64_SOURCE` or `-D_GNU_SOURCE` in C library compilation
3. that user then chose to call `mmap` and not `mmap64`.

And here, we see that `-D_GNU_SOURCE` is a relatively common compile-time define (it's on by default in `g++`).

So, I guess we are back to just using `Native.LONG_SIZE` and providing the user the option to specify 64 bit, and providing the "safe" wrapper functions for `mmap` and `ftruncate` that will attempt the 64-bit version with a 64-bit arg, to maximize portability.